### PR TITLE
add baichuan/chatglm2 +lora+agent examples

### PR DESCRIPTION
**Pull request by Jintao-Huang**

### CODE REVIEW

File: examples/pytorch/llm_agent/_common.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
+import os
+import random
+import re
+import sys
+import math
+import json
+import ast
+import datetime as dt
+from typing import List, Tuple, Dict, Callable, Optional, Union, Any, Literal
```
Comment (tastelikefeet) [inline examples/pytorch/llm_agent/_common.py:9]: ImportError: cannot import name 'Literal'

Comment (tastelikefeet) [reply examples/pytorch/llm_agent/_common.py:9]: python 3.9

File: examples/pytorch/llm_agent/_common.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    if drop_last:
+        T_max = dataset_len // batch_size
+    else:
+        T_max = math.ceil(dataset_len / batch_size)
+    T_max *= max_epochs
+    return T_max
+# 
+
+def tokenize_function(system: str, user: str, assistant: Optional[str], tokenizer) -> Dict[str, Any]:
+    """只适用baichuan和chatglm2. 其他模型需进行test"""
```
Comment (tastelikefeet) [inline examples/pytorch/llm_agent/_common.py:135]: Only En comments pls

File: examples/pytorch/llm_agent/_common.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    system_text = SYSTEM_TEXT.format(system=system)
+    user_text = USER_TEXT.format(user=user)
+    system_text_ids: List[int] = tokenizer(system_text, return_attention_mask=False, 
+                                          add_special_tokens=True)["input_ids"]
+    user_text_ids: List[int] = tokenizer(user_text, return_attention_mask=False, 
+                                          add_special_tokens=False)["input_ids"]
+    assistant_p_input_ids: List[int] = tokenizer(ASSISTANT_PROMPT, return_attention_mask=False, 
+                                          add_special_tokens=False)["input_ids"]
+
+    # tokenizer.bos_token_id: 避免assistant为空
```
Comment (tastelikefeet) [inline examples/pytorch/llm_agent/_common.py:145]: Same here

File: examples/pytorch/llm_agent/_common.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    for i, (n, p) in enumerate(named_p):
+        if i >= max_lines:
+            logger.info("...")
+            break
+        logger.info(f"{n}: requires_grad={p.requires_grad}")
+
+@METRICS.register_module(group_key=default_group, module_name='my_metric')
+class MyMetric(Metric):
+    def __init__(self, vocab_size: int):
+        self.acc = Accuracy("multiclass", num_classes=vocab_size)
```
Comment (tastelikefeet) [inline examples/pytorch/llm_agent/_common.py:241]: Why acc, not Rouge or Bleu?

File: examples/pytorch/llm_agent/_common.py (hunk 1 above)
Comment (Jintao-Huang) [reply examples/pytorch/llm_agent/_common.py:9]: python >= 3.8 supports Literal
https://peps.python.org/pep-0586/

File: examples/pytorch/llm_agent/_common.py (hunk 2 above)
Comment (Jintao-Huang) [reply examples/pytorch/llm_agent/_common.py:135]: OK

File: examples/pytorch/llm_agent/_common.py (hunk 4 above)
Comment (Jintao-Huang) [reply examples/pytorch/llm_agent/_common.py:244]: The rouge bleu based approach requires a token-by-token generation approach, which leads to long validation times. 
acc is used here for simple validation.

**Review by tastelikefeet — APPROVED** [review]

**Pull request merged**